public/app/features/alerting/unified/utils/notification-policies.ts

```diff
@@ -0,0 +1,105 @@
+import { AlertmanagerGroup, MatcherOperator, ObjectMatcher, Route } from 'app/plugins/datasource/alertmanager/types';
+
+import { normalizeMatchers } from './amroutes';
+
+export type Label = [string, string];
+type OperatorPredicate = (labelValue: string, matcherValue: string) => boolean;
+
+const OperatorFunctions: Record<MatcherOperator, OperatorPredicate> = {
+  [MatcherOperator.equal]: (lv, mv) => lv === mv,
+  [MatcherOperator.notEqual]: (lv, mv) => lv !== mv,
+  [MatcherOperator.regex]: (lv, mv) => Boolean(lv.match(new RegExp(mv))),
+  [MatcherOperator.notRegex]: (lv, mv) => !Boolean(lv.match(new RegExp(mv))),
+};
+
+function isLabelMatch(matcher: ObjectMatcher, label: Label) {
+  const [labelKey, labelValue] = label;
+  const [matcherKey, operator, matcherValue] = matcher;
+
+  // not interested, keys don't match
+  if (labelKey !== matcherKey) {
+    return false;
+  }
+
+  const matchFunction = OperatorFunctions[operator];
+  if (!matchFunction) {
+    throw new Error(`no such operator: ${operator}`);
+  }
+
+  return matchFunction(labelValue, matcherValue);
+}
+
+// check if every matcher returns "true" for the set of labels
+function matchLabels(matchers: ObjectMatcher[], labels: Label[]) {
+  return matchers.every((matcher) => {
+    return labels.some((label) => isLabelMatch(matcher, label));
+  });
+}
+
+// Match does a depth-first left-to-right search through the route tree
+// and returns the matching routing nodes.
+function findMatchingRoutes<T extends Route>(root: T, labels: Label[]): T[] {
```

Comment:
As we discussed yesterday, I think having a map with labels pointing to the list of the nodes in this tree would be a good improvement for this findMatching performance.

Reply:
Yea this one bothers me a lot, it's incredibly slow for large trees and a large amount of alert groups. I'm considering dropping this feature for now and include it behind a feature flag so we have some time to optimize it without halting the release of this new policies view.